jormungandr/src/utils/async_msg.rs

```diff
@@ -61,6 +73,51 @@ impl<Msg> Sink for MessageBox<Msg> {
     }
 }
 
+/// State for asynchronous sending of a message over a `MessageBox`
+/// that can be driven as a standalone task.
+pub struct SendTask<Msg> {
+    mbox: MessageBox<Msg>,
+    pending: Option<Msg>,
+    logger: Logger,
+}
+
+impl<Msg> SendTask<Msg> {
+    fn handle_mbox_error<T>(&self, err: SendError<T>) {
+        error!(
+            self.logger,
+            "failed to enqueue message for processing";
+            "reason" => %err,
+        )
+    }
+}
+
+impl<Msg> Future for SendTask<Msg> {
+    type Item = ();
+    type Error = ();
+
+    fn poll(&mut self) -> Poll<(), ()> {
+        loop {
+            if self.pending.is_some() {
+                let msg = self.pending.take().unwrap();
+                let async_sink = self
+                    .mbox
+                    .start_send(msg)
+                    .map_err(|e| self.handle_mbox_error(e))?;
+                if let AsyncSink::NotReady(msg) = async_sink {
+                    self.pending = Some(msg);
+                    return Ok(Async::NotReady);
+                }
+            } else {
+                try_ready!(self
+                    .mbox
+                    .poll_complete()
+                    .map_err(|e| self.handle_mbox_error(e)));
+                return Ok(().into());
+            }
+        }
+    }
```

Comment:
I wonder if this could be reduced to `mbox.clone().send().map_err(<log it>)`?

Reply:
It has to be a named type, but a closure in `.map_err(...)` does not currently allow that as far as I know.

Reply:
I mean the whole `SendTask` structure with custom `poll` implementation. How is it better than gluing together a few built-in `Stream` wrappers?

Reply:
Oh, the problem is that it must be stored in a struct field, so it needs a concrete, named type.